scipy/signal/_filter_design.py

```diff
@@ -1428,100 +1464,133 @@ def zpk2sos(z, p, k, pairing='nearest'):
     # 4. Further optimizations of the section ordering / pole-zero pairing.
     # See the wiki for other potential issues.
 
-    valid_pairings = ['nearest', 'keep_odd']
+    if pairing is None:
+        pairing = 'minimal' if analog else 'nearest'
+
+    valid_pairings = ['nearest', 'keep_odd', 'minimal']
     if pairing not in valid_pairings:
         raise ValueError('pairing must be one of %s, not %s'
                          % (valid_pairings, pairing))
-    if len(z) == len(p) == 0:
-        return array([[k, 0., 0., 1., 0., 0.]])
 
-    # ensure we have the same number of poles and zeros, and make copies
-    p = np.concatenate((p, np.zeros(max(len(z) - len(p), 0))))
-    z = np.concatenate((z, np.zeros(max(len(p) - len(z), 0))))
-    n_sections = (max(len(p), len(z)) + 1) // 2
-    sos = zeros((n_sections, 6))
+    if analog and pairing != 'minimal':
+        raise ValueError('for analog zpk2sos conversion, '
+                         + 'pairing must be "minimal"')
+
+    if len(z) == len(p) == 0:
+        if not analog:
+            return np.array([[k, 0., 0., 1., 0., 0.]])
+        else:
+            return np.array([[0., 0., k, 0., 0., 1.]])
+
+    if pairing != 'minimal':
+        # ensure we have the same number of poles and zeros, and make copies
+        p = np.concatenate((p, np.zeros(max(len(z) - len(p), 0))))
+        z = np.concatenate((z, np.zeros(max(len(p) - len(z), 0))))
+        n_sections = (max(len(p), len(z)) + 1) // 2
+
+        if len(p) % 2 == 1 and pairing == 'nearest':
+            p = np.concatenate((p, [0.]))
+            z = np.concatenate((z, [0.]))
+        assert len(p) == len(z)
```

Comment:
Can you update this (and below)? `assert` cannot be used for control flow. I am not even sure it's needed in this case.

Reply:
This was in the original code, I've just indented it for the `if`. It's not being used for control-flow; it's just an ordinary assert.  It's an error here if `len(p)!=len(z)`. I don't mind removing it.

Reply:
I know, it's a nice to have here. If you don't I will see if we have an issue about assert or create one. Because things like this are easy to forget.

Reply:
I tend to leave "harmless" code alone; there's also a `del p, z` towards the end of `zpk2sos` that I've left in; I don't think that helps, and unlike the assertion it barely documents anything (other than an intent to no longer use p or z - but the function is about to end!).

Reply:
I personally don't mind if we clean up things as we go. In any case we need to review the code so if we can ship trivial things in the batch, why not, everybody is happy (no need to review twice the same code) and we don't leave things that we all identified that could have been improved (we try to reduce the churn to contribute, but for trivial things, the entry barrier is still there).

I will resolve this as it's again optional.